### PR TITLE
Some MAX_SIZE macros are too small when PSA ECC is accelerated

### LINKED ISSUE #6622: Some MAX_SIZE macros are too small when PSA ECC is accelerated
At some time before Mbed TLS 2.28 and 3.0, we added support for signature through PSA accelerators. Accelerators can support elliptic curves that the core doesn't. So the [definition of `PSA_VENDOR_ECC_MAX_CURVE_BITS`](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-2.28.0/include/psa/crypto_sizes.h#L203), which determines `PSA_EXPORT_KEY_PAIR_MAX_SIZE`, `PSA_EXPORT_PUBLIC_KEY_MAX_SIZE` and PSA_SIGNATURE_MAX_SIZE, is wrong. If the largest accelerated curve is larger than the largest built-in curve (for example if ECC is accelerated only), the `MAX_SIZE` macros may end up being wrong (mostly, if there is no RSA support).

This will also concern `PSA_RAW_KEY_AGREEMENT_OUTPUT_MAX_SIZE` when we [add support for key agreement acceleration](https://github.com/Mbed-TLS/mbedtls/pull/6426#discussion_r1018015270).

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

The goal of this PR is to fix `PSA_VENDOR_ECC_MAX_CURVE_BITS` when some of the curves are accelerated.
Resolve #6622 

~This PR is a draft because even though ECDH is accelerated there are tests failing due to key derivation. [This issue is known](https://github.com/Mbed-TLS/mbedtls/issues/6839#issuecomment-1365821625) and should be addressed in a different PR.~

~This PR depends on #7321 ([explanation here](https://github.com/Mbed-TLS/mbedtls/pull/7103#issuecomment-1482268750))~

## Gatekeeper checklist

- [x] **changelog** required as this is a bug fix, provided
- [x] **backport** not required - the bug is only visible when using new features from 3.x (driver-only ECC)
- [x] **tests** provided

### CODE REVIEW

File: include/psa/crypto_sizes.h [hunk 1]
```diff
@@ -192,31 +192,31 @@
 
 /* The maximum size of an ECC key on this implementation, in bits.
  * This is a vendor-specific macro. */
-#if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED)
+#if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED) || defined(PSA_WANT_ECC_SECP_R1_521)
```
Comment (ronald-cron-arm) [inline include/psa/crypto_sizes.h:195]:
```suggestion
#if defined(PSA_WANT_ECC_SECP_R1_521)
```
config_psa.h takes care of defining MBEDTLS_ECP_DP_SECP521R1_ENABLED if PSA_WANT_ECC_SECP_R1_521 is defined and vice versa.

Comment (valeriosetti) [reply include/psa/crypto_sizes.h:195]: Isn't `MBEDTLS_ECP_DP_SECP521R1_ENABLED` re-defined if `MBEDTLS_PSA_ACCEL_ECC_SECP_R1_521` is not (i.e. if that curve is not accelerated)?
```
#if defined(PSA_WANT_ECC_SECP_R1_521)
#if !defined(MBEDTLS_PSA_ACCEL_ECC_SECP_R1_521)
```

The same hold for the others

Comment (ronald-cron-arm) [reply include/psa/crypto_sizes.h:195]: > Isn't `MBEDTLS_ECP_DP_SECP521R1_ENABLED` re-defined if `MBEDTLS_PSA_ACCEL_ECC_SECP_R1_521` is not (i.e. if that curve is not accelerated)?

If MBEDTLS_ECP_DP_SECP521R1_ENABLED is already defined in mbedtls_config.h, yes it is re-defined in config_psa.h if MBEDTLS_PSA_CRYPTO_CONFIG and PSA_WANT_ECC_SECP_R1_521 are enabled and MBEDTLS_PSA_ACCEL_ECC_SECP_R1_521 is not. As long as the definitions are the same the compilers do not complain.

Comment (ronald-cron-arm) [reply include/psa/crypto_sizes.h:195]: > But the curves are accelerated in this test... Am I missing something?

I am not sure what you mean here. To come back to my change suggestion, if the curves are accelerated it means that we are using the PSA configuration (MBEDTLS_PSA_CRYPTO_CONFIG is enabled) and thus the support for the curve is determined by the definition or not of PSA_WANT_ECC_SECP_R1_521.

Otherwise, if we are not using the PSA configuration (MBEDTLS_PSA_CRYPTO_CONFIG is disabled), the support for the curve is to start with determined by the definition or not of MBEDTLS_ECP_DP_SECP521R1_ENABLED in mbedtls_config.h and in config_psa.h we have:

#if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED)                                   
#define MBEDTLS_PSA_BUILTIN_ECC_SECP_R1_521 1                                   
#define PSA_WANT_ECC_SECP_R1_521                                                
#endif

Thus in all cases the curve ECC_SECP_R1_521 is supported if and only if PSA_WANT_ECC_SECP_R1_521 is defined and we do not need to care about MBEDTLS_ECP_DP_SECP521R1_ENABLED here.

Comment (valeriosetti) [reply include/psa/crypto_sizes.h:195]: > We do not support an external PSA library

Ok, I think this is the bit of information I was missing. I thought that the purpose of `libtestdriver1.a` was to create a sort of generic PSA compliant driver library to interface to (and with "generic" I mean that I could be replaced with any other PSA compliant implementation).

> If MBEDTLS_ECP_DP_xxx is accessible through the PSA API, then crypto_config.h will take care of defining PSA_WANT_ECC_xxx

But having `MBEDTLS_ECP_DP_xxx` doesn't cause some extra code (for example from `ecp_curves.c`) to be taken into the build even though that curve is accelerated?

Comment (gilles-peskine-arm) [reply include/psa/crypto_sizes.h:195]: The purpose of libtestdriver1 is to test the driver interface in the core (the part that isn't in libtestdriver1): we want to validate that when a driver for some mechanism is enabled, the core calls the driver (and not the built-in implementation).

If `MBEDTLS_ECP_DP_xxx` is enabled, that causes the relevant part of `ecp.c` and `ecp_curves.c` to be built (generic code + any curve-specific optimized code). If the curve is accelerated, and you don't need access to that curve from legacy functions that can't yet dispatch to PSA, then you don't need `MBEDTLS_ECP_DP_xxx` — assuming that Mbed TLS behaves correctly if a curve is only built as accelerated, which is what we're working on here.

**Comment by valeriosetti** [comment]
Note for reviewers: I had to set legacy `MBEDTLS_ECP_DP_xxx` requirements on some key derivations (see [this commit](https://github.com/Mbed-TLS/mbedtls/pull/7103/commits/f29e760cf990497cdf8bf92ff677725614bef3b4)) because:
- driver's dispatching is not implemented yet
- when deriving weierstrass key it goes through ECP module, which itself uses `MBEDTLS_ECP_DP_xxx` definitions, which are however excluded from the library part. 

BTW the double dependency `PSA_WANT_xxx` + `MBEDTLS_ECP_DP_xxx` seems reasonable to me until key derivation does not implement driver's dispatching. @mpg Wdyt?

File: include/psa/crypto_sizes.h (hunk 1 above)
Comment (mpg) [reply include/psa/crypto_sizes.h:195]: In addition to what Ronald said, I think it's important to note that here we're "in PSA context": this is a PSA header and we're defining a PSA macro. So as a general principle we should use only `PSA_WANT` macros. (This is a pre-existing issue, but since we're fixing this part of the code, let's do it the right way.)

Then, code that's doing ECC using the PSA APIs can just use this macro. Code that's doing ECC and might be using either PSA or legacy will need to check both `PSA_VENDOR_ECC_MAX_CURVE_BITS` and `MBEDTLS_ECP_MAX_BITS` and probably take the maxium or whatever, and that's OK.

But for pure-PSA code, things should remain clean.

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -187,7 +187,7 @@ pre_initialize_variables () {
 
     # CFLAGS and LDFLAGS for Asan builds that don't use CMake
     # default to -O2, use -Ox _after_ this if you want another level
-    ASAN_CFLAGS='-O2 -Werror -fsanitize=address,undefined -fno-sanitize-recover=all'
+    ASAN_CFLAGS='-O0 -g -Werror -fsanitize=address,undefined -fno-sanitize-recover=all'
```
Comment (mpg) [inline tests/scripts/all.sh:190]: Looks like leftover from debugging.

File: tests/suites/test_suite_psa_crypto.data [hunk 3]
```diff
@@ -6118,7 +6118,7 @@ depends_on:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256:PSA_WANT_ALG_TLS12_PRF
 derive_key_exercise:PSA_ALG_TLS12_PRF(PSA_ALG_SHA_256):"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":"000102030405060708090a0b0c":"f0f1f2f3f4f5f6f7f8f9":PSA_KEY_TYPE_DERIVE:400:PSA_KEY_USAGE_DERIVE:PSA_ALG_HKDF(PSA_ALG_SHA_256)
 
 PSA key derivation: HKDF-SHA-256 -> ECC secp256r1, exercise ECDSA
-depends_on:PSA_WANT_ALG_ECDSA:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256
+depends_on:PSA_WANT_ALG_ECDSA:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256:MBEDTLS_ECP_DP_SECP256R1
```
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.data:6121]: Minor: it's equivalent in practice, but as a matter of principle I think this should be `MBEDTLS_PSA_BUILTIN_ECC_SECP_R1_256`. In the long run we want to completely eliminate `MBEDTLS_` macros from the PSA space, other than perhaps `MBEDTLS_PSA_` macros.

Again, in this case the two names are equivalent, and the dependency is correct (until we implemented driver support for ECC key derivation) but using the `MBEDTLS_PSA_` macro makes future automated checking easier, when we work on eliminating `MBEDTLS_` dependencies that, unlike that one, are not correct.

**Review by mpg — COMMENTED** [review]
Just a first pass.

File: tests/scripts/all.sh [hunk 4]
```diff
@@ -2479,6 +2479,105 @@ component_test_psa_crypto_config_accel_ecc () {
     make test
 }
 
+component_test_verify_ec_max_size () {
```
Comment (mpg) [inline tests/scripts/all.sh:2482]: Naming nit: I'd prefer something in the `component_test_psa_crypto_config_accel_` space. Perhaps something like `component_test_psa_crypto_config_accel_ecc_curves_except_p192` and then the other one coming later could be `accel_ecc_curves_except_x25519`? (P-192 is NIST's name for secp192r1, picked it here because it's shorter.) Or `accel_all_curves_except_xxx`?

File: tests/scripts/all.sh [hunk 5]
```diff
@@ -2479,6 +2479,105 @@ component_test_psa_crypto_config_accel_ecc () {
...
+    msg "test: MBEDTLS_PSA_CRYPTO_CONFIG + accel EC algs + USE_PSA_CRYPTO"
```
Comment (mpg) [inline tests/scripts/all.sh:2483]: I think this message as well as the one below need adapting.

File: tests/scripts/all.sh [hunk 6]
```diff
@@ -2479,6 +2479,105 @@ component_test_psa_crypto_config_accel_ecc () {
...
+    scripts/config.py set MBEDTLS_PSA_CRYPTO_DRIVERS
+    scripts/config.py set MBEDTLS_PSA_CRYPTO_CONFIG
+    scripts/config.py set MBEDTLS_USE_PSA_CRYPTO
+
+    # disable modules for which we have drivers
+    scripts/config.py unset MBEDTLS_ECDSA_C
+    scripts/config.py unset MBEDTLS_ECDH_C
+    scripts/config.py unset MBEDTLS_ECJPAKE_C
+
+    # disable key exchanges dependencies from the modules above
```
Comment (mpg) [inline tests/scripts/all.sh:2522]: That should no longer be necessary after all the work in the other PRs. OTOH, after disabling RSA we'll want to disable RSA-based key exchanges.

Comment (valeriosetti) [reply tests/scripts/all.sh:2522]: Right, disabling of ECDSA/ECDH/ECJPAKE key exchanges is a leftover of when I started this PR. OTOH RSA ones should be removed, you're right.

File: tests/suites/test_suite_psa_crypto.data (hunk 3 above)
Comment (valeriosetti) [reply tests/suites/test_suite_psa_crypto.data:6121]: I didn't think about this detail, but I totally agree! :)

**Comment by valeriosetti** [comment]
@mpg I just set the `need-preceding-pr` label on this PR because I forgot to mention that it depends on #7321 for a [fix that will be introduced](https://github.com/Mbed-TLS/mbedtls/pull/7103/commits/3cc9ccdae2264873f48fc57268d2da115b30fbda) by the top level ECC epic; #7321 in particular.

File: tests/scripts/all.sh [hunk 7]
```diff
@@ -2468,6 +2468,103 @@ component_test_psa_crypto_config_accel_ecc () {
...
+    loc_accel_flags="$loc_accel_flags -DLIBTESTDRIVER1_MBEDTLS_PSA_ACCEL_ALG_SHA_224"
+    loc_accel_flags="$loc_accel_flags -DLIBTESTDRIVER1_MBEDTLS_PSA_ACCEL_ALG_SHA_256"
+    loc_accel_flags="$loc_accel_flags -DLIBTESTDRIVER1_MBEDTLS_PSA_ACCEL_ALG_SHA_384"
+    loc_accel_flags="$loc_accel_flags -DLIBTESTDRIVER1_MBEDTLS_PSA_ACCEL_ALG_SHA_512"
+    make -C tests libtestdriver1.a CFLAGS="$ASAN_CFLAGS $loc_accel_flags" LDFLAGS="$ASAN_CFLAGS"
+
+    # Configure and build the main libraries
+    # ---------------------------------------
+
+    # start with default + driver support + USE_PSA_CRYPTO
```
Comment (mpg) [inline tests/scripts/all.sh:2501]: I think we should start with full here, to make ensure proper coverage. Actually, this component should be very similar to the one added in #7321, except this one has RSA disabled and most curves accelerated.

File: tests/scripts/all.sh [hunk 8]
```diff
@@ -2468,6 +2468,103 @@ component_test_psa_crypto_config_accel_ecc () {
...
+    loc_accel_list="$loc_accel_list ECC_SECP_R1_224"
+    loc_accel_list="$loc_accel_list ECC_SECP_K1_224"
+    loc_accel_list="$loc_accel_list ECC_SECP_R1_192"
+    loc_accel_list="$loc_accel_list ECC_SECP_K1_192"
+
+    # build and link with test drivers
+    loc_accel_flags="$loc_accel_flags $( echo "$loc_accel_list" | sed 's/[^ ]* */-DMBEDTLS_PSA_ACCEL_&/g' )"
+    make CFLAGS="$ASAN_CFLAGS -I../tests/include -I../tests -I../../tests -DPSA_CRYPTO_DRIVER_TEST -DMBEDTLS_TEST_LIBTESTDRIVER1 $loc_accel_flags" LDFLAGS="-ltestdriver1 $ASAN_CFLAGS"
+
+    # make sure excluded modules were not auto-re-enabled by accident
```
Comment (mpg) [inline tests/scripts/all.sh:2557]: We could probably go a bit further here and add something like
```
# The only built-in curve is Short Weierstrass, so ECP shouldn't have support for Montgomery curves.
# Functions with mxz in their name are specific to Montgomery curves.
not grep mxz library/ecp.o
```
This is a bit weak in that all functions with mxz in their name are static so there's no guarantee they will be visible in the `.o` file, but I tried and one of them seems to be visible at `-O2` with GCC (but not Clang). Anyway, I don't think adding that hurts, as the worst that can happen is it tests nothing.

If/when we do another component with Curve25519 as the only curve, the similar test will be more robust, as we have public function that's only defined when there are Short Weierstrass curves: `mbedtls_ecp_muladd` so we can grep for that and be certain that the result means something.

File: tests/scripts/all.sh (hunk 7 above)
Comment (valeriosetti) [reply tests/scripts/all.sh:2501]: This is something I noticed, but I couldn't find an elegant way to compress the 2 without ending up with something unreadable :/

Comment (mpg) [reply tests/scripts/all.sh:2501]: Yes, I've been thinking about it too and came to the same conclusion :)

File: tests/scripts/all.sh (hunk 8 above)
Comment (valeriosetti) [reply tests/scripts/all.sh:2557]: Update = Indeed I just found 1 test that was depending on `MBEDTLS_ECP_DP_SECP192R1_ENABLED` and that didn't build with this new configuration. So it was worth to add the new test case.

Comment (mpg) [reply tests/scripts/all.sh:2570]: > I'm not convinced that is better to add the test for Curve25519 now

Typo for "I'm now convinced"? Anyway, I see you added it and I agree it's a good idea, thanks!

File: include/mbedtls/check_config.h [hunk 9]
```diff
@@ -359,7 +359,8 @@
 
 #if defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED) &&                    \
     ( !defined(MBEDTLS_PK_HAVE_JPAKE) ||                                    \
-      !defined(MBEDTLS_ECP_DP_SECP256R1_ENABLED) )
+      !(defined(MBEDTLS_ECP_DP_SECP256R1_ENABLED) || \
```
Comment (mpg) [inline include/mbedtls/check_config.h:362]: I'm afraid it's a bit more complex: TLS will only take advantage of the presence of the curve in PSA is `USE_PSA` is defined. So, I think this should be
```c
!((!defined(MBEDTLS_USE_PSA_CRYPTO) && defined(MBEDTLS_ECP_DP_SECP256R1_ENABLED)) || \
  (defined(MBEDTLS_ECP_DP_SECP256R1_ENABLED) && defined(PSA_WANT_ECC_SECP_R1_256)))
```
Obviously, if we find ourselved using this expression too often (and perhaps once is already too often for readability?) we could define a macro `MBEDTLS_PK_HAVE_CURVE_xxx` like we did for ECDH and the others.

File: tests/scripts/all.sh [hunk 10]
```diff
@@ -2498,10 +2498,16 @@ component_test_psa_crypto_config_accel_all_curves_except_p192 () {
     # Configure and build the main libraries
     # ---------------------------------------
 
-    # start with default + driver support + USE_PSA_CRYPTO
-    scripts/config.py set MBEDTLS_PSA_CRYPTO_DRIVERS
+    # full config + USE_PSA_CRYPTO
```
Comment (mpg) [inline tests/scripts/all.sh:2507]: That would be "full config (includes USE_PSA_CRYPTO)" I think. (Or perhaps "includes USE_PSA, TLS 1.3 and driver support" even.)

File: tests/scripts/all.sh [hunk 11]
```diff
@@ -2527,9 +2533,7 @@ component_test_psa_crypto_config_accel_all_curves_except_p192 () {
     # Explicitly disable all SW implementation for elliptic curves and enable
     # their accelerated version (this excludes the built-in automatic inclusion
     # from "config_psa.h")
-    # Just leave SW implementation for the curve with the smallest bit size
-    # (MBEDTLS_ECP_DP_SECP192R1_ENABLED) for allowing to build with ECP_C.
-    # scripts/config.py unset MBEDTLS_ECP_DP_SECP192R1_ENABLED
+    scripts/config.py unset MBEDTLS_ECP_DP_SECP192R1_ENABLED
```
Comment (mpg) [inline tests/scripts/all.sh:2536]: Minor: the compete enumeration is a bit tedious, perhaps we should use a for loop as done in `component_test_everest_curve25519_only()`?

Perhaps the same trick can be done with `ECC_xxx` symbols by sed-ing `include/psa/crypto_config.h` while at it.

File: tests/scripts/all.sh [hunk 12]
```diff
@@ -2555,6 +2559,9 @@ component_test_psa_crypto_config_accel_all_curves_except_p192 () {
     loc_accel_list="$loc_accel_list ECC_SECP_K1_224"
     loc_accel_list="$loc_accel_list ECC_SECP_R1_192"
     loc_accel_list="$loc_accel_list ECC_SECP_K1_192"
+    # Just leave SW implementation for the curve with the smallest bit size
+    # (MBEDTLS_ECP_DP_SECP192R1_ENABLED) for allowing to build with ECP_C.
```
Comment (mpg) [inline tests/scripts/all.sh:2563]: I think this comment needs updating now that the curve is passed as an argument to this function.

File: tests/scripts/all.sh [hunk 13]
```diff
@@ -2564,13 +2571,28 @@ component_test_psa_crypto_config_accel_all_curves_except_p192 () {
     not grep mbedtls_ecdh_ library/ecdh.o
     not grep mbedtls_ecdsa_ library/ecdsa.o
     not grep mbedtls_ecjpake_ library/ecjpake.o
+    if [ $BUILTIN_CURVE == "MBEDTLS_ECP_DP_SECP192R1_ENABLED" ]; then
+        # The only built-in curve is Short Weierstrass, so ECP shouldn't have support for Montgomery curves.
+        # Functions with mxz in their name are specific to Montgomery curves.
+        not grep mxz library/ecp.o
+    else
```
Comment (mpg) [inline tests/scripts/all.sh:2578]: This assume that if it's not `MBEDTLS_ECP_DP_SECP192R1_ENABLED` then it's `CURVE25519`, which I think should be either tested or documented. Also, please add a comment as above saying that `ecp_muladd` is specific to Short Weierstrass curves.

File: tests/scripts/all.sh [hunk 14]
```diff
@@ -2468,6 +2468,131 @@ component_test_psa_crypto_config_accel_ecc () {
...
+# - component_test_psa_crypto_config_accel_all_curves_except_x25519
+# to build and test with all accelerated curves a part from the specified one.
+psa_crypto_config_accel_all_curves_except_one () {
+    BUILTIN_CURVE=$1
+
+    msg "build: PSA_CRYPTO_CONFIG + all accelerated EC algs (excl secp192r1) + USE_PSA_CRYPTO"
+
+    # Accelerate all EC algs (all EC curves are automatically accelerated as
+    # well in the built-in version due to the "PSA_WANT_xxx" symbols in
+    # "crypto_config.h")
```
Comment (mpg) [inline tests/scripts/all.sh:2482]: Minor: we could get rid of this parenthesis by just adding the curve heres, which seems a bit more logical to me: the top-level list is for things that we want to be present in libtestdriver and accelerated in the main libraries, which is the case for curves here. Either 
```
   loc_accel_list="ALG_ECDH \
                    ALG_ECDSA ALG_DETERMINISTIC_ECDSA \
                    ALG_JPAKE \
                    KEY_TYPE_ECC_KEY_PAIR KEY_TYPE_ECC_PUBLIC_KEY \
                    ECC_BRAINPOOL_P_R1_256 ECC_BRAINPOOL_P_R1_384 ECC_BRAINPOOL_P_R1_512 \
                    ECC_MONTGOMERY_255 ECC_MONTGOMERY_448 \
                    ECC_SECP_K1_192 ECC_SECP_K1_224 ECC_SECP_K1_256 \
                    ECC_SECP_R1_192 ECC_SECP_R1_224 ECC_SECP_R1_256 ECC_SECP_R1_384 ECC_SECP_R1_521"
```
or using sed and a for loop to enumerate and add the curves, as discussed in another comment.

Wdyt?

**Review by mpg — CHANGES_REQUESTED** [review]
Looking pretty good, and thanks for adding the 2nd component! Other than the EC J-PAKE depenency, the only things I found are minor nits.

**Comment by mpg** [comment]
> **backport** -> question for gatekeepers: since this is a bugfix, should this be backported?

I don't think so. I agree it's technically a bug, but one that only shows in configurations that we don't really support in 2.28, considering that even the possibility of building with driver-only ECDSA or ECDH is new in development. Other reviewers please shout if you disagree.
 
> *tests -> question for gatekeepers: should I test it with outcome-analysis.sh to be sure I did not change the number of tests executed?

I don't think that's needed here, as we didn't massively change existing dependency declarations.

File: tests/scripts/all.sh (hunk 8 above)
Comment (valeriosetti) [reply tests/scripts/all.sh:2570]: Ops, sorry, I wrote that comment in a rush and that's the result :D

File: tests/scripts/all.sh (hunk 11 above)
Comment (valeriosetti) [reply tests/scripts/all.sh:2536]: > Perhaps the same trick can be done with ECC_xxx symbols by sed-ing include/psa/crypto_config.h while at it.

Ok for the 1st part, but this one is not clear to me. Don't we want `PSA_WANT_ECC_` symbols when building the library? I thought that `PSA_WANT_ECC_` + `MBEDTLS_PSA_ACCEL_` combo on the main library side is what we wanted. Am I wrong?

Comment (mpg) [reply tests/scripts/all.sh:2536]: Yes, I meant just to get the list of symbols, then add them to `loc_accel_list` - not unset them with `scripts/config.py` indeed! Something like:
```
for CURVE in $( sed ... ); do
   loc_accel_list="$loc_accel_list $CURVE"
done
```

Comment (valeriosetti) [reply tests/scripts/all.sh:2536]: Ok, I see, I misunderstood the previous comment then. I got it now ;)

Comment (mpg) [reply tests/scripts/all.sh:2536]: Actually now that I think of it, it's not just cosmetic, getting the list from the config file would also be more future-proof as we might add new curves, there are at least two Edwards curves that are serious candidates if we can ever get to it...

File: tests/scripts/all.sh (hunk 14 above)
Comment (valeriosetti) [reply tests/scripts/all.sh:2482]: Actually, when building the driver we want to have the built-in version of all the curves supported (since there is no other lower lever SW which can provide them), so according to `config_psa.h` we want:

- `PSA_WANT_ECC_` and NOT `MBEDTLS_PSA_ACCEL_ECC_` --> `MBEDTLS_PSA_BUILTIN_ECC_`

In this case the `crypto_config_test_driver_extensions.h` do no perform any "adaptation" for the curves' defines.

In short: I would keep the current declaration for curves where it is now. Is it wrong?

**Review by mpg — DISMISSED** [review]
LGTM, unfortunately it grew a conflict in the meantime.

Note to self: the version I'm happy with is ffc3f22c3c3bf979e522873b5031d6efc0009f2f

**Comment by mpg** [comment]
PS: we can probably wait for #7321 to be merged before resolving the conflict, in order to minimize the number of times this is rebased.

**Comment by mpg** [comment]
With #7321 merged, this can now be rebased. (Note to self: I'm happy with aaae30421c60903e31a18ed92468f6496786b363)

File: tests/scripts/all.sh [hunk 15]
```diff
@@ -2285,6 +2288,176 @@ component_test_psa_crypto_config_reference_all_ec_algs_use_psa () {
     tests/ssl-opt.sh
 }
 
+component_test_psa_crypto_config_accel_ecc () {
```
Comment (mpg) [inline tests/scripts/all.sh:2291]: This was removed in the previous PR, so I think re-introducing it here was a rebasing accident.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks almost all good to me, except for the accidental re-introduction of component that was previously removed.

**Review by mpg — APPROVED** [review]
LGTM

File: tests/scripts/all.sh [hunk 16]
```diff
@@ -2285,6 +2288,118 @@ component_test_psa_crypto_config_reference_all_ec_algs_use_psa () {
...
+    scripts/config.py unset MBEDTLS_ECP_RESTARTABLE
+
+    # disable modules for which we have drivers
+    scripts/config.py unset MBEDTLS_ECDSA_C
+    scripts/config.py unset MBEDTLS_ECDH_C
+    scripts/config.py unset MBEDTLS_ECJPAKE_C
+
+    # Ensure also RSA_C is disabled so that the size of the public/private
+    # keys cannot be taken from there
+    scripts/config.py unset MBEDTLS_RSA_C
```
Comment (ronald-cron-arm) [inline tests/scripts/all.sh:2345]: As the PSA configuration is enabled, some PSA_WANT_ALG_RSA_xxx are defined without acceleration this does not have any impact on the build I think.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
I only have one minor comment otherwise this looks good to me. Thanks for the two non-regression all.sh components. I have checked that when reverting the changes in crypto_size.h the PSA crypto unit tests actually fail in the two components.

**Review by mpg — APPROVED** [review]
LGTM

**Comment by mpg** [comment]
Despite this being a bug, and sort-of present in 2.28 already, I'm inclined not to backport, because the bug is only visible when using driver-only ECC, which only became possible in 3.4/development. (Practically, in 2.28 we'd have no way to test the fix like we did here.) @valeriosetti @ronald-cron-arm please shout if you disagree.

**Comment by valeriosetti** [comment]
> I'm inclined not to backport, because the bug is only visible when using driver-only ECC, which only became possible in 3.4/development

I agree with you. Do I need to write this somewhere (i.e. the fact that we are intentionally not backporting this fix) for future reference?

**Review by mpg — APPROVED** [review]
LGTM

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, thanks for the changes.

**Comment by ronald-cron-arm** [comment]
> Despite this being a bug, and sort-of present in 2.28 already, I'm inclined not to backport, because the bug is only visible when using driver-only ECC, which only became possible in 3.4/development. (Practically, in 2.28 we'd have no way to test the fix like we did here.) @valeriosetti @ronald-cron-arm please shout if you disagree.

I am fine with that.
